docs/core/howto/defer.rst

```diff
@@ -456,8 +457,75 @@ Now if someone calls ``cancel()`` on the ``Deferred`` returned from ``HTTPClient
 Care should be taken not to ``callback()`` a Deferred that has already been cancelled.
 
 
+.. _core-howto-defer-deferreds-timeouts:
+
+Timeouts
+--------
+
+Timeouts are a special case of :ref:`Cancellation <core-howto-defer-deferreds-cancellation>`.
+Let's say we have a :api:`twisted.internet.defer.Deferred <Deferred>` representing a task that may take a long time.
+We want to put an upper bound on that task, so we want the :api:`twisted.internet.defer.Deferred <Deferred>` to time
+out X seconds in the future.
+
+A convenient API to do so is :api:`twisted.internet.defer.Deferred.addTimeout <Deferred.addTimeout>`.
+By default, it will fail with a :api:`twisted.internet.defer.TimeoutError <TimeoutError>` if the :api:`twisted.internet.defer.Deferred <Deferred>` hasn't fired (with either an errback or a callback) within ``timeout`` seconds.
+
+.. code-block:: python
+
+    import random
+    from twisted.internet import task
+
+    def f():
+        return "Hopefully this will be called in 3 seconds or less"
+
+    def main(reactor):
+        delay = random.uniform(1, 5)
+
+        def called(result):
+            print("{0} seconds later:".format(delay), result)
+
+        d = task.deferLater(reactor, delay, f)
+        d.addTimeout(3, reactor).addBoth(called)
+
+        return d
+
+    # f() will be timed out if the random delay is greater than 3 seconds
+    task.react(main)
+
+
+:api:`twisted.internet.defer.Deferred.addTimeout <Deferred.addTimeout>` uses the :api:`twisted.internet.defer.Deferred.cancel <Deferred.cancel>` function under the hood, but can distinguish between a user's call to :api:`twisted.internet.defer.Deferred.cancel <Deferred.cancel>` and a cancellation due to a timeout.
+By default, :api:`twisted.internet.defer.Deferred.addTimeout <Deferred.addTimeout>` translates a :api:`twisted.internet.defer.CancelledError <CancelledError>` produced by the timeout into a :api:`twisted.internet.error.TimeoutError <TimeoutError>`.
+
+However, if you provided a custom :ref:`cancellation <core-howto-defer-deferreds-cancellation>` when creating the :api:`twisted.internet.defer.Deferred <Deferred>`, then cancelling it may not produce a :api:`twisted.internet.defer.CancelledError <CancelledError>`.  In this case, the default behavior of :api:`twisted.internet.task.timeoutDeferred <timeoutDeferred>` is to preserve whatever callback or errback value your custom cancellation function produced.  This can be useful if, for instance, a cancellation or timeout should produce a default value instead of an error.
```

Comment:
Please change api link to `addTimeout` in "In this case, the default behavior of..." sentence.